Add raw Zephyr SystemLayer implementation (noSelect)

ArekBalysNordic opened the pull request:
#### Summary

The new Zephyr System Layer implementation does not use Zephyr's BSD socket API (select). This lightweight option is suitable when the Matter network layer interacts directly with the networking stack, instead of going through Zephyr's NET L2 Layer. As a result, it does not require sockets or the select API to function. This implementation serves as an alternative to SystemLayerSelect, and both can be used alongside Zephyr if needed.

Additionally, I added the capability to run the nRFConnect platform using the OpenThread TCP/UDP stack directly instead of Zephyr sockets. Thanks to this, we can test the new Zephyr System layer implementation that does not use sockets in the nRFConnect samples. I've defined OpenThread native init params in all-clusters-app.

#### Testing

Tested manually on the nRFConnect all-clusters-app example.
Enabled this new implementation by default in all-clusters-app, so its build process is tested in the nRFConnect workflow.

Steps to test the new solution locally:

1. Build the all-clusters-app example for the nrfconnect nrf52840 platform using the following command:
```
west build -p -b nrf52840dk/nrf52840 -- -DCONFIG_CHIP_USE_OPENTHREAD_ENDPOINT=y
```
2. Test example as usual, the new Zephyr System Layer without socket will be used.

adigie committed:
[nrfconnect] Add configuration for CHIP_USE_OPEN_THREAD_ENDPOINT

The new configuration allows enabling and disabling using
OpenThread endpoints as well as using Zephyr SystemLayer
implementation without Sockets.

Signed-off-by: Adrian Gielniewski <adrian.gielniewski@nordicsemi.no>
Signed-off-by: Arkadiusz Balys <arkadiusz.balys@nordicsemi.no>

adigie committed:
[Zephyr] Add raw Zephyr SystemLayer implementation (noSelect)

The new Zephyr System Layer implementation does not use Zephyr's
BSD socket API (select). This lightweight option is suitable when
the Matter network layer interacts directly with the networking
stack, instead of going through Zephyr's NET L2 Layer. As a result,
it does not require sockets or the select API to function.

This implementation serves as an alternative to SystemLayerSelect,
and both can be used alongside Zephyr if needed.

Signed-off-by: Adrian Gielniewski <adrian.gielniewski@nordicsemi.no>
Co-authored-by: Arkadiusz Balys <arkadiusz.balys@nordicsemi.no>

ArekBalysNordic committed:
[nrfconnect] Enable OpenThread endpoint in all-clusters-app

Enabled OpenThread endpoint instead of zephyr's socket in
nRFConnect all-clusters-app.

Signed-off-by: Arkadiusz Balys <arkadiusz.balys@nordicsemi.no>

Damian-Nordic reviewed (commented):
A couple initial comments.

ArekBalysNordic commented on src/platform/nrfconnect/ConnectivityManagerImpl.cpp:
```diff
@@ -17,16 +17,28 @@
+using UDPEndPointImpl = chip::Inet::UDPEndPointImplOT;
+#elif CHIP_SYSTEM_CONFIG_USE_POSIX_SOCKETS || CHIP_SYSTEM_CONFIG_USE_ZEPHYR_SOCKETS
 #include <inet/UDPEndPointImplSockets.h>
+using UDPEndPointImpl = chip::Inet::UDPEndPointImplSockets;
```
That's right, I've refactored it.

ArekBalysNordic commented on src/platform/nrfconnect/ConnectivityManagerImpl.cpp:
```diff
@@ -17,16 +17,28 @@
 #ifndef CONFIG_ARCH_POSIX
 #include <zephyr/net/net_if.h>
 #endif
+#if CHIP_DEVICE_CONFIG_ENABLE_THREAD
```
Right, not needed, removed.

Damian-Nordic commented on config/nrfconnect/chip-module/Kconfig:
```diff
@@ -366,4 +366,11 @@ config CHIP_ENABLE_BDX_LOG_TRANSFER
 	  Enables the BDX protocol for diagnostics log transfer purposes.
 
+config CHIP_USE_OPENTHREAD_ENDPOINT
+  bool "Use OpenThread TCP/UDP stack directly"
```
Fix indentation (use TAB instead of spaces)

Damian-Nordic commented on src/include/platform/internal/GenericPlatformManagerImpl_Zephyr.ipp:
```diff
@@ -109,14 +114,36 @@ void GenericPlatformManagerImpl_Zephyr<ImplClass>::_UnlockChipStack(void)
     mShouldRunEventLoop = false;
+
+#if !CHIP_SYSTEM_CONFIG_USE_SOCKETS
+    // If the platform timer is being updated by a thread other than the event loop thread,
```
This comment seems copy-pasted but is not relevant for this method (stopping the CHIP thread).

Damian-Nordic commented on src/include/platform/internal/GenericPlatformManagerImpl_Zephyr.ipp:
```diff
@@ -194,10 +226,74 @@ void GenericPlatformManagerImpl_Zephyr<ImplClass>::_RunEventLoop(void)
+
+        // Call into the system layer to dispatch the callback functions for all timers
+        // that have expired.
+        CHIP_ERROR err = systemLayer.HandlePlatformTimer();
```
Note that you call this every time `k_msgq_get` exits, even if the function has exited because of an event, not timeout. This makes some calls to `HandlePlatformTimer` unnecessary. This could be avoided by moving this to line 269 above.

Damian-Nordic commented on src/system/SystemLayerImplZephyr.cpp:
```diff
@@ -0,0 +1,235 @@
+
+CHIP_ERROR LayerImplZephyr::TriggerPlatformTimerUpdate()
+{
+    VerifyOrReturnError(IsInitialized(), CHIP_ERROR_INCORRECT_STATE);
```
This is private method that is only used in contexts where this is already checked, seems unnecessary.

Damian-Nordic commented on src/system/SystemLayerImplZephyr.cpp:
```diff
@@ -0,0 +1,235 @@
+CHIP_ERROR LayerImplZephyr::TriggerPlatformTimerUpdate()
+{
+    VerifyOrReturnError(IsInitialized(), CHIP_ERROR_INCORRECT_STATE);
+    CHIP_ERROR status = PlatformEventing::StartTimer(*this, Clock::kZero);
```
Callers of this function do not check the return value. Perhaps, make it `void` and just log the error if we can't handle errors gracefully?

Damian-Nordic commented on src/system/SystemLayerImplZephyr.cpp:
```diff
@@ -0,0 +1,235 @@
+    Clock::Timeout remainingTime = mTimerList.GetRemainingTime(onComplete, appState);
+    if (remainingTime.count() < delay.count())
+    {
+        if (remainingTime == Clock::kZero)
```
nit: I know this is copied but I wonder if this is needed - `StartTimer` calls `CancelTimer`, which seems to check `mExpiredTimers` as well...

Damian-Nordic commented on src/system/SystemLayerImplZephyr.cpp:
```diff
@@ -0,0 +1,235 @@
+    VerifyOrReturn(timer != nullptr);
+    mTimerPool.Release(timer);
+
+    TriggerPlatformTimerUpdate();
```
This seems unnecessary - even if the CHIP thread waits for N seconds and then the timer is already gone, this is fine and we don't have to wait it sooner to figure this out. Especially that the timer may then be rescheduled and wake the thread again.

Damian-Nordic reviewed (approved):
A few suggestions to consider but looks good overall

kkasperczyk-no commented on config/nrfconnect/chip-module/Kconfig:
```diff
@@ -366,4 +366,11 @@ config CHIP_ENABLE_BDX_LOG_TRANSFER
 
+config CHIP_USE_OPENTHREAD_ENDPOINT
+  bool "Use OpenThread TCP/UDP stack directly"
+  depends on OPENTHREAD && !WIFI_NRF70
```
How about replacing `WIFI_NRF70` with `WIFI`? I don't think that using specific board/family is the best idea here. Otherwise we will have to update it in the future.

kkasperczyk-no commented on examples/all-clusters-app/nrfconnect/Kconfig:
```diff
@@ -33,6 +33,9 @@ choice OPENTHREAD_DEVICE_TYPE
 endchoice
 
+config CHIP_USE_OPENTHREAD_ENDPOINT
+	default y
```
Don't we want to enable it for the whole nrfconnect Thread configuration?

ArekBalysNordic replied:
Maybe for now, let's test both configurations.

ArekBalysNordic replied:
Okay, to avoid making this PR too large and complicated to review, I will handle it in the next PR. In this case, let's focus on the new Zephyr System Layer implementation.

ArekBalysNordic committed:
Code review

Signed-off-by: Arkadiusz Balys <arkadiusz.balys@nordicsemi.no>

andy31415 commented on src/include/platform/internal/GenericPlatformManagerImpl_Zephyr.ipp:
```diff
@@ -180,6 +195,20 @@ void GenericPlatformManagerImpl_Zephyr<ImplClass>::_RunEventLoop(void)
     }
     mShouldRunEventLoop = true;
 
+#if CHIP_SYSTEM_CONFIG_USE_SOCKETS
```
These if/else/endif are scattered throughout the code.

Would it be reasonably clean to create a separte IPP file one for ZephyrSockets and one ZephyrNoSockets ? That way we have one if for includes and don't split out the code as much.

andy31415 commented on src/system/SystemLayerImplZephyr.h:
```diff
@@ -0,0 +1,73 @@
+
+/**
+ *    @file
+ *      This file declares an implementation of LayerImplZephyr.
```
delete this comment

andy31415 commented on src/system/SystemLayerImplZephyr.h:
```diff
@@ -0,0 +1,73 @@
+    CHIP_ERROR ScheduleWork(TimerCompleteCallback onComplete, void * appState) override;
+
+public:
+    // Platform implementation.
```
I am unsure what to do with this comment ... what does it mean? is it related to GetNextAwakenTime? Could you refine or delete if there is nothing to say?

ArekBalysNordic replied:
I will remove it.

ArekBalysNordic committed:
Separate GenericPlatformManagerImpl_Zephyr.ipp files

Separated the Zephyr implementation of GenericPlatformManager
to version with and without using Select().

Signed-off-by: Arkadiusz Balys <arkadiusz.balys@nordicsemi.no>

ArekBalysNordic commented on src/system/SystemLayerImplZephyr.cpp:
```diff
@@ -0,0 +1,229 @@
+
+/**
+ *    @file
+ *      This file implements LayerImplZephyr.
```
Ok, I agree with you. I thought it was something required, but right, it's actually a waste of storage bytes. Let me remove it.
